### PR TITLE
Trying forward declarations of needed curand functions

### PR DESCRIPTION
**Pull request by pjh5**
Right now libcaffe2_gpu.so links against both libcurand_static.a and THCTensorRandom.cu.o , which when built statically includes the definitions for some functions that are also in libcurand. This leads to a multiple definitions error.

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
Blugh, but the approach looks sound.

If you want to generalize this a little, refactor out the forward declarations into a standalone header, so anyone else who wants to include `curand_mtgp32_host.h` can include that header instead; i.e., this is a generally useful forward declaration that all use-sites of this header want.

File: aten/src/THC/THCTensorRandom.cu [hunk 1]
```diff
@@ -6,12 +6,35 @@
...
 #include <curand.h>
 #include <curand_kernel.h>
+
+// Forward declarations of functions that are defined in libcurand_static.a
+// This is to avoid multiple-definitions of these when statically linking
+// cudarand in both Caffe2 and ATen
+#if CAFFE2_STATIC_LINK_CUDA()
+curandStatus_t curandMakeMTGP32Constants(const mtgp32_params_fast_t params[],mtgp32_kernel_params_t * p);
+void mtgp32_init_state(unsigned int state[],
+              const mtgp32_params_fast_t *para, unsigned int seed);
```
Comment (yinghai) [inline aten/src/THC/THCTensorRandom.cu]: weird indent?

Comment (pjh5) [reply aten/src/THC/THCTensorRandom.cu]: Oh right. I was going to change that but forgot

**Comment by pjh5** [comment]
@ezyang where should the new header go? aten/src/THC seems like the wrong place to me

**Comment by pjh5** [comment]
I will use this branch for temporary testing, do son't merge it yet.

**Pull request closed without merging**